### PR TITLE
TypeScript support

### PR DESCRIPTION
**Pull request by thibaudcolas**
This adds TypeScript support (parser, compatible rules and needed plugins, documentation) directly at the configuration level. Previously we would have recommended TypeScript projects add this as extra configuration from the base configuration.

The goal here is for TypeScript support to be completely transparent – particularly from the perspective of projects that _don’t_ use TypeScript.

This is available to test as a pre-release:

```bash
npm install --save-dev eslint-config-torchbox@next
```

The README details how to then use this.

---

- [x] Thibaud to finish PR
- [x] James review approach – design decisions & usage as a dev on a project
- [x] Thibaud publish pre-release
- [x] James Trial on Next.js careers site
- [x] Thibaud Trial on Wagtail
- [x] Thibaud PR to wagtail-kit

### CODE REVIEW

**Comment by thibaudcolas** [comment]
- Thibaud to finish this
- Assign to James to review approach – design decisions & usage as a dev on a project
- Thibaud publish pre-release
- James Trial on Next.js careers site
- Thibaud Trial on Wagtail
- Thibaud PR to wagtail-kit

**Comment by thibaudcolas** [comment]
Pre-release: https://www.npmjs.com/package/eslint-config-torchbox/v/0.6.0-0

To install:

```
npm install --save-dev eslint-config-torchbox@next
```

ESLint config:

```
module.exports = {
    // See https://github.com/torchbox/eslint-config-torchbox for rules.
    extends: 'torchbox/typescript',
    env: {
        browser: true,
        commonjs: true,
        es6: true,
        jest: true,
        node: true,
    },
};
```

File: typescript.js [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                'constructor-super': 'off',
+                'getter-return': 'off',
+                'no-const-assign': 'off',
+                'no-dupe-args': 'off',
+                'no-dupe-class-members': 'off',
+                'no-dupe-keys': 'off',
+                'no-func-assign': 'off',
+                'no-new-symbol': 'off',
+                'no-obj-calls': 'off',
+                'no-redeclare': 'off',
```
Comment (jhancock532) [inline typescript.js:318]: In line 149 we deactivate `no-redeclare` then define the typescript rule. Is this override necessary?

**Review by jhancock532 — COMMENTED** [review]
Some potentially duplicated rule overrides, otherwise this setup looks great.

File: src/dependencies.test.js [hunk 2]
```diff
@@ -4,12 +4,18 @@ const pkg = require('../package.json');
...
     });
 
     it('has stable peerDependencies', () => {
         expect(pkg.peerDependencies).toMatchInlineSnapshot(`
             Object {
+              "@typescript-eslint/eslint-plugin": "^5.0.0",
+              "@typescript-eslint/parser": "^5.0.0",
               "eslint": "^8.2.0",
               "eslint-config-airbnb": "^19.0.2",
               "eslint-config-prettier": "^8.3.0",
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply src/dependencies.test.js:21]: Yes, I think this helps quite a bit for projects where updating dependencies would be a big deal (in this case linting rules would change).

File: typescript.js [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+                'react/require-default-props': 0,
+                // Copy from eslint-config-airbnb-typescript
+                // https://github.com/iamturns/eslint-config-airbnb-typescript/blob/d30b7d2a41f3f2cf899cad3ed69e3b7eacbc7234/lib/shared.js#L256
+                // The following rules are enabled in Airbnb config, but are already checked (more thoroughly) by the TypeScript compiler
+                // Some of the rules also fail in TypeScript files, for example: https://github.com/typescript-eslint/typescript-eslint/issues/662#issuecomment-507081586
+                'constructor-super': 'off',
+                'getter-return': 'off',
+                'no-const-assign': 'off',
+                'no-dupe-args': 'off',
+                'no-dupe-class-members': 'off',
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply typescript.js:313]: Indeed, nice catch. This is copy-pasted from eslint-config-airbnb-typescript, so we should report it there as well. Do you want to do it or should I?

File: typescript.js (hunk 1 above)
Comment (thibaudcolas) [reply typescript.js:318]: Likely not! I’ll edit, I think we should also report this to eslint-config-airbnb-typescript.

**Pull request merged**